acl/policy.go

```diff
@@ -243,3 +309,501 @@ func Parse(rules string, sentinel sentinel.Evaluator) (*Policy, error) {
+		}
+		if sp.Intentions != "" && !isPolicyValid(sp.Intentions) {
+			return nil, fmt.Errorf("Invalid service intentions policy: %#v", sp)
+		}
+		if err := isSentinelValid(sentinel, sp.Policy, sp.Sentinel); err != nil {
+			return nil, fmt.Errorf("Invalid service Sentinel policy: %#v, got error:%v", sp, err)
+		}
+
+		p.ServicePrefixes = append(p.ServicePrefixes, sp)
+	}
+
+	// Validate the session policies
+	for _, sp := range lp.Sessions {
+		if !isPolicyValid(sp.Policy) {
+			return nil, fmt.Errorf("Invalid session policy: %#v", sp)
+		}
+
+		p.SessionPrefixes = append(p.SessionPrefixes, sp)
+	}
+
+	// Validate the user event policies
+	for _, ep := range lp.Events {
+		if !isPolicyValid(ep.Policy) {
+			return nil, fmt.Errorf("Invalid event policy: %#v", ep)
+		}
+
+		p.EventPrefixes = append(p.EventPrefixes, ep)
+	}
+
+	// Validate the prepared query policies
+	for _, pq := range lp.PreparedQueries {
+		if !isPolicyValid(pq.Policy) {
+			return nil, fmt.Errorf("Invalid query policy: %#v", pq)
+		}
+
+		p.PreparedQueryPrefixes = append(p.PreparedQueryPrefixes, pq)
+	}
+
+	// Validate the keyring policy - this one is allowed to be empty
+	if lp.Keyring != "" && !isPolicyValid(lp.Keyring) {
+		return nil, fmt.Errorf("Invalid keyring policy: %#v", lp.Keyring)
+	} else {
+		p.Keyring = lp.Keyring
+	}
+
+	// Validate the operator policy - this one is allowed to be empty
+	if lp.Operator != "" && !isPolicyValid(lp.Operator) {
+		return nil, fmt.Errorf("Invalid operator policy: %#v", lp.Operator)
+	} else {
+		p.Operator = lp.Operator
+	}
+
+	return p, nil
+}
+
+// NewPolicyFromSource is used to parse the specified ACL rules into an
+// intermediary set of policies, before being compiled into
+// the ACL
+func NewPolicyFromSource(id string, revision uint64, rules string, syntax SyntaxVersion, sentinel sentinel.Evaluator) (*Policy, error) {
+	if rules == "" {
+		// Hot path for empty source
+		return &Policy{ID: id, Revision: revision}, nil
+	}
+
+	var policy *Policy
+	var err error
+	switch syntax {
+	case SyntaxLegacy:
+		policy, err = parseLegacy(rules, sentinel)
+	case SyntaxCurrent:
+		policy, err = parseCurrent(rules, sentinel)
+	default:
+		return nil, fmt.Errorf("Invalid rules version: %d", syntax)
+	}
+
+	if err == nil {
+		policy.ID = id
+		policy.Revision = revision
+	}
+	return policy, err
+}
+
+// takesPrecedenceOver returns true when permission a
+// should take precedence over permission b
+func takesPrecedenceOver(a, b string) bool {
+	if a == PolicyDeny {
+		return true
+	} else if b == PolicyDeny {
+		return false
+	}
+
+	if a == PolicyWrite {
+		return true
+	} else if b == PolicyWrite {
+		return false
+	}
+
+	if a == PolicyList {
+		return true
+	} else if b == PolicyList {
+		return false
+	}
+
+	if a == PolicyRead {
+		return true
+	} else if b == PolicyRead {
+		return false
+	}
+
+	return false
+}
+
+func multiPolicyID(policies []*Policy) uint64 {
+	var id uint64
+	hasher := fnv.New64a()
```

Comment:
FNV is _not_ a crypto hash function. I assume this is OK but maybe we should add a comment here to state that it's never expected that the ID returned be used as a secret?

If the individual hash IDs are random with a crypto source then this might be OK in practice, but we should still use a crypto hash if we ever expect the output hash to be used as a secret/unguessable/unforgeable value.

Reply:
Hmm, originally I was using fnv because I wasn't going to use it for crypto purposes. Except I just realized that it is being exported as a field for read only token operations that don't currently expose the secret and therefore should probably be a crypto hash.

This individual case is not a problem but there are others that could be.